### PR TITLE
Add plot function for SpaceNet datasets

### PR DESCRIPTION
**Pull request by ashnair1**
Changes:
- Add plot function to SpaceNet base class
- Explicitly check whether raster CRS is equal to vector CRS and transform polygon features accordingly
- ~Add `contrast_stretch` to correctly convert between types~
- Added axis param to `percentile_normalization` and updated utils ` __all__`

### CODE REVIEW

File: torchgeo/datasets/spacenet.py [hunk 1]
```diff
@@ -271,6 +289,53 @@ def _download(self, collections: List[str], api_key: Optional[str] = None) -> No
...
+    ) -> Figure:
+        """Plot a sample from the dataset.
+
+        Args:
+            sample: a sample returned by :meth:`__getitem__`
+            show_titles: flag indicating whether to show titles above each panel
+            suptitle: optional string to use as a suptitle
+
+        Returns:
+            a matplotlib Figure with the rendered sample
```
Comment (adamjstewart) [inline torchgeo/datasets/spacenet.py:306]: Can you add:
```rst
.. versionadded:: 0.2
```
so we know when this function was added to the library?

File: torchgeo/datasets/utils.py [hunk 2]
```diff
@@ -482,3 +483,30 @@ def rgb_to_mask(
         if isinstance(cmask, np.ndarray):
             mask[cmask.all(axis=-1)] = i
     return mask
+
+
+def contrast_stretch(
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py:488]: @calebrob6 this looks similar to the function you are adding in #251 but with `lower=0, upper=100`. Maybe we should wait on #251 and use that function here?

**Comment by ashnair1** [comment]
Plot sample

From SpaceNet2, PAN image and label
sn2_pan

From SpaceNet7
sn7_train

File: torchgeo/datasets/spacenet.py (hunk 1 above)
Comment (ashnair1) [reply torchgeo/datasets/spacenet.py:306]: Done

File: torchgeo/datasets/spacenet.py [hunk 3]
```diff
@@ -271,6 +289,55 @@ def _download(self, collections: List[str], api_key: Optional[str] = None) -> No
...
+        fig, axs = plt.subplots(ncols=ncols, figsize=(ncols * 8, 8))
+        if not isinstance(axs, np.ndarray):
+            axs = [axs]
+        axs[0].imshow(image)
+        axs[0].axis("off")
+        if show_titles:
+            axs[0].set_title("Image")
+
+        if show_mask:
+            axs[1].imshow(mask)
```
Comment (calebrob6) [inline torchgeo/datasets/spacenet.py:332]: does this need a `vmin` and `vmax` to behave correctly in the case where not all classes are present in an image?

Comment (calebrob6) [reply torchgeo/datasets/spacenet.py:332]: you could probably add `interpolation="none"` so you don't get this effect:

image

Comment (ashnair1) [reply torchgeo/datasets/spacenet.py:332]: > you could probably add `interpolation="none"` so you don't get this effect:
> 
> image

Setting `axs[1].imshow(mask, interpolation=None)` doesn't seem to have an effect on the plot

Comment (ashnair1) [reply torchgeo/datasets/spacenet.py:332]: > does this need a `vmin` and `vmax` to behave correctly in the case where not all classes are present in an image?

All the SpaceNet datasets (except SpaceNet5 IIRC) have one class (buildings or roads).

**Comment by ashnair1** [comment]
Visualisation note:
When `percentile normalization` was done bandwise (before 5ea1c2126) there was a one to one correspondence between the plot torchgeo produces and the plot that QGIS (default params i.e stretch to minmax, 2% - 98%) shows. 

Overall normalisation
def1


Bandwise and how QGIS shows it by default
def2

File: torchgeo/datasets/utils.py (hunk 2 above)
Comment (ashnair1) [reply torchgeo/datasets/utils.py]: Removed `contrast_stretch` and switched to `percentile normalization`

**Comment by calebrob6** [comment]
Strange! One of the S2 datasets I was working on looked very strange when I did bandwise min-max normalization so I just changed it to global and it looked fine.

Do you think we should add an `axis` parameter to `percentile_normalization` that gets forwarded through to numpy?

**Comment by ashnair1** [comment]
Forwarding `axis` param seems like a good idea. Gives the user a way to control the type of normalization.

The SpaceNet images that I've shown above are from Worldview 2 so it could be a sensor thing (dtype). I usually plot it to be similar to QGIS default params but that's just my preference.

**Comment by calebrob6** [comment]
Sounds good to me too! Do you want to do that?

**Comment by ashnair1** [comment]
Sure. Would it be better to do that as a separate commit or shall I just add it here?

**Comment by calebrob6** [comment]
Yeah go for it!!

**Comment by ashnair1** [comment]
Added axis param to `percentile_normalization` and added it to `__all__`

File: torchgeo/datasets/utils.py [hunk 4]
```diff
@@ -485,7 +486,10 @@ def rgb_to_mask(
 
 
 def percentile_normalization(
-    img: np.ndarray, lower: float = 2, upper: float = 98  # type: ignore[type-arg]
+    img: np.ndarray,  # type: ignore[type-arg]
+    lower: float = 2,
+    upper: float = 98,
+    axis: Optional[Union[int, Tuple[int, int]]] = None,
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py:492]:
```suggestion
    axis: Optional[Union[int, Sequence[int]]] = None,
```
I think both lists and tuples of any length are valid.

File: torchgeo/datasets/utils.py [hunk 5]
```diff
@@ -497,15 +501,21 @@ def percentile_normalization(
         img: image to normalize
         lower: lower percentile in range [0,100]
         upper: upper percentile in range [0,100]
+        axis: Axis or axes along which the percentiles are computed. The default
+        is to compute the percentile(s) along a flattened version of the array.
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py:505]:
```suggestion
            is to compute the percentile(s) along a flattened version of the array.
```

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**